File: jdeep/integration/edr_connector.py
"""
EDR Connector Module
====================

Provides connectors for various EDR/AV platforms.
Supports CrowdStrike, Microsoft Defender ATP, Carbon Black, and more.
"""

import logging
import subprocess
from typing import Dict, Any, List, Optional
from abc import ABC, abstractmethod

logger = logging.getLogger(__name__)


class BaseEDRConnector(ABC):
    """Abstract base class for EDR connectors."""
    
    @abstractmethod
    def connect(self, credentials: Dict[str, Any]) -> bool:
        """Connect to the EDR platform."""
        pass
    
    @abstractmethod
    def get_status(self) -> Dict[str, Any]:
        """Get EDR status."""
        pass
    
    @abstractmethod
    def get_alerts(self, time_range: int) -> List[Dict[str, Any]]:
        """Get alerts from the EDR."""
        pass


class CrowdStrikeConnector(BaseEDRConnector):
    """CrowdStrike Falcon connector."""
    
    def __init__(self):
        self.client_id: Optional[str] = None
        self.client_secret: Optional[str] = None
        self.base_url: str = "https://api.crowdstrike.com"
        self.token: Optional[str] = None
        self.connected: bool = False
    
    def connect(self, credentials: Dict[str, Any]) -> bool:
        """
        Connect to CrowdStrike Falcon.
        
        Args:
            credentials: Dict with client_id and client_secret
        """
        self.client_id = credentials.get("client_id")
        self.client_secret = credentials.get("client_secret")
        
        if not self.client_id or not self.client_secret:
            logger.error("CrowdStrike credentials missing")
            return False
        
        try:
            # Authenticate with OAuth2
            import requests
            
            response = requests.post(
                f"{self.base_url}/oauth2/token",
                data={
                    "client_id": self.client_id,
                    "client_secret": self.client_secret,
                },
                headers={"Content-Type": "application/x-www-form-urlencoded"}
            )
            
            if response.status_code == 201:
                self.token = response.json().get("access_token")
                self.connected = True
                logger.info("Connected to CrowdStrike Falcon")
                return True
            else:
                logger.error(f"CrowdStrike auth failed: {response.status_code}")
                return False
                
        except ImportError:
            logger.error("requests library not installed")
            return False
        except Exception as e:
            logger.error(f"CrowdStrike connection error: {e}")
            return False
    
    def get_status(self) -> Dict[str, Any]:
        """Get CrowdStrike sensor status."""
        return {
            "name": "CrowdStrike Falcon",
            "connected": self.connected,
            "running": self._check_sensor_running(),
        }
    
    def get_alerts(self, time_range: int = 3600) -> List[Dict[str, Any]]:
        """Get recent CrowdStrike alerts."""
        if not self.connected:
            return []
        
        try:
            import requests
            
            response = requests.get(
                f"{self.base_url}/alerts/queries/alerts/v1",
                headers={"Authorization": f"Bearer {self.token}"},
                params={"filter": f"created_timestamp:>now-{time_range}s"}
            )
            
            if response.status_code == 200:
                return response.json().get("resources", [])
            return []
            
        except Exception as e:
            logger.error(f"Error fetching CrowdStrike alerts: {e}")
            return []
    
    def _check_sensor_running(self) -> bool:
        """Check if CrowdStrike sensor is running."""
        try:
            result = subprocess.run(
                ["sc", "query", "csfalconservice"],
                capture_output=True,
                text=True
            )
            return "RUNNING" in result.stdout
        except:
            return False


class DefenderConnector(BaseEDRConnector):
    """Microsoft Defender ATP connector."""
    
    def __init__(self):
        self.tenant_id: Optional[str] = None
        self.client_id: Optional[str] = None
        self.client_secret: Optional[str] = None
        self.token: Optional[str] = None
        self.connected: bool = False
    
    def connect(self, credentials: Dict[str, Any]) -> bool:
        """
        Connect to Microsoft Defender ATP.
        
        Args:
            credentials: Dict with tenant_id, client_id, client_secret
        """
        self.tenant_id = credentials.get("tenant_id")
        self.client_id = credentials.get("client_id")
        self.client_secret = credentials.get("client_secret")
        
        if not all([self.tenant_id, self.client_id, self.client_secret]):
            logger.error("Defender ATP credentials missing")
            return False
        
        try:
            import requests
            
            response = requests.post(
                f"https://login.microsoftonline.com/{self.tenant_id}/oauth2/v2.0/token",
                data={
                    "client_id": self.client_id,
                    "client_secret": self.client_secret,
                    "scope": "https://api.securitycenter.microsoft.com/.default",
                    "grant_type": "client_credentials"
                }
            )
            
            if response.status_code == 200:
                self.token = response.json().get("access_token")
                self.connected = True
                logger.info("Connected to Microsoft Defender ATP")
                return True
            else:
                logger.error(f"Defender ATP auth failed: {response.status_code}")
                return False
                
        except Exception as e:
            logger.error(f"Defender ATP connection error: {e}")
            return False
    
    def get_status(self) -> Dict[str, Any]:
        """Get Defender status."""
        status = {
            "name": "Microsoft Defender ATP",
            "connected": self.connected,
            "running": False,
            "real_time_protection": False,
        }
        
        try:
            result = subprocess.run(
                ["powershell", "-Command", 
                 "Get-MpComputerStatus | Select-Object RealTimeProtectionEnabled, AntivirusEnabled"],
                capture_output=True,
                text=True
            )
            
            status["running"] = "True" in result.stdout
            status["real_time_protection"] = "RealTimeProtectionEnabled" in result.stdout and "True" in result.stdout
            
        except Exception as e:
            logger.debug(f"Error checking Defender status: {e}")
        
        return status
    
    def get_alerts(self, time_range: int = 3600) -> List[Dict[str, Any]]:
        """Get recent Defender alerts."""
        alerts = []
        
        try:
            # Get local Defender threat detections
            result = subprocess.run(
                ["powershell", "-Command",
                 f"Get-MpThreatDetection | Where-Object {{$_.InitialDetectionTime -gt (Get-Date).AddSeconds(-{time_range})}}"],
                capture_output=True,
                text=True
            )
            
            if result.stdout.strip():
                alerts.append({
                    "source": "Windows Defender",
                    "type": "ThreatDetection",
                    "details": result.stdout.strip()
                })
                
        except Exception as e:
            logger.debug(f"Error fetching Defender alerts: {e}")
        
        return alerts


class CarbonBlackConnector(BaseEDRConnector):
    """VMware Carbon Black connector."""
    
    def __init__(self):
        self.api_url: Optional[str] = None
        self.api_key: Optional[str] = None
        self.org_key: Optional[str] = None
        self.connected: bool = False
    
    def connect(self, credentials: Dict[str, Any]) -> bool:
        """Connect to Carbon Black."""
        self.api_url = credentials.get("api_url")
        self.api_key = credentials.get("api_key")
        self.org_key = credentials.get("org_key")
        
        if not all([self.api_url, self.api_key, self.org_key]):
            logger.error("Carbon Black credentials missing")
            return False
        
        # Validate connection
        try:
            import requests
            
            response = requests.get(
                f"{self.api_url}/appservices/v6/orgs/{self.org_key}/alerts",
                headers={
                    "X-Auth-Token": self.api_key,
                    "Content-Type": "application/json"
                }
            )
            
            self.connected = response.status_code == 200
            if self.connected:
                logger.info("Connected to Carbon Black")
            return self.connected
            
        except Exception as e:
            logger.error(f"Carbon Black connection error: {e}")
            return False
    
    def get_status(self) -> Dict[str, Any]:
        """Get Carbon Black status."""
        return {
            "name": "Carbon Black",
            "connected": self.connected,
            "running": self._check_sensor_running(),
        }
    
    def get_alerts(self, time_range: int = 3600) -> List[Dict[str, Any]]:
        """Get recent Carbon Black alerts."""
        if not self.connected:
            return []
        
        try:
            import requests
            
            response = requests.get(
                f"{self.api_url}/appservices/v6/orgs/{self.org_key}/alerts",
                headers={
                    "X-Auth-Token": self.api_key,
                    "Content-Type": "application/json"
                },
                params={"time_range": f"-{time_range}s"}
            )
            
            if response.status_code == 200:
                return response.json().get("results", [])
            return []
            
        except Exception as e:
            logger.error(f"Error fetching Carbon Black alerts: {e}")
            return []
    
    def _check_sensor_running(self) -> bool:
        """Check if Carbon Black sensor is running."""
        try:
            result = subprocess.run(
                ["sc", "query", "CbDefense"],
                capture_output=True,
                text=True
            )
            return "RUNNING" in result.stdout
        except:
            return False


class EDRConnector:
    """
    Main EDR connector class.
    
    Manages multiple EDR platform connections.
    """
    
    CONNECTORS = {
        "crowdstrike": CrowdStrikeConnector,
        "defender": DefenderConnector,
        "carbonblack": CarbonBlackConnector,
    }
    
    def __init__(self, config=None):
        """
        Initialize EDR Connector.
        
        Args:
            config: ConfigManager instance
        """
        self.config = config
        self.connectors: Dict[str, BaseEDRConnector] = {}
        
        logger.info("EDRConnector initialized")
    
    def connect(self, edr_name: str, credentials: Dict[str, Any]) -> bool:
        """
        Connect to an EDR platform.
        
        Args:
            edr_name: Name of EDR platform
            credentials: API credentials
            
        Returns:
            True if successful
        """
        edr_name = edr_name.lower()
        
        if edr_name not in self.CONNECTORS:
            logger.error(f"Unknown EDR: {edr_name}")
            return False
        
        connector = self.CONNECTORS[edr_name]()
        if connector.connect(credentials):
            self.connectors[edr_name] = connector
            return True
        
        return False
    
    def get_all_status(self) -> List[Dict[str, Any]]:
        """Get status of all EDR systems."""
        status_list = []
        
        # Check connected EDRs
        for name, connector in self.connectors.items():
            status_list.append(connector.get_status())
        
        # Check local EDR products even if not connected via API
        defender = DefenderConnector()
        defender_status = defender.get_status()
        if defender_status not in [s["name"] for s in status_list]:
            status_list.append(defender_status)
        
        return status_list
    
    def get_alerts(
        self,
        edr_name: Optional[str] = None,
        time_range: int = 3600
    ) -> List[Dict[str, Any]]:
        """
        Get alerts from EDR platforms.
        
        Args:
            edr_name: Specific EDR or None for all
            time_range: Time range in seconds
            
        Returns:
            List of alerts
        """
        alerts = []
        
        if edr_name:
            connector = self.connectors.get(edr_name.lower())
            if connector:
                alerts.extend(connector.get_alerts(time_range))
        else:
            for connector in self.connectors.values():
                alerts.extend(connector.get_alerts(time_range))
        
        return alerts

File: src/integration/edr_connector.cpp
/*
 * EDR Connectors - Implementation
 * =================================
 * Author: Jdeep
 */

#include "integration/integration_manager.hpp"
#include <iostream>

#ifdef _WIN32
#include <windows.h>
#include <comdef.h>
#include <wbemidl.h>
#endif

namespace edr {
namespace integration {

// ============================================================================
// CROWDSTRIKE CONNECTOR
// ============================================================================

CrowdStrikeConnector::CrowdStrikeConnector(const std::string& clientId,
                                            const std::string& clientSecret)
    : clientId_(clientId)
    , clientSecret_(clientSecret)
    , connected_(false) {
}

CrowdStrikeConnector::~CrowdStrikeConnector() {
    disconnect();
}

bool CrowdStrikeConnector::connect() {
    std::cout << "[*] Connecting to CrowdStrike Falcon API..." << std::endl;
    
    // OAuth2 authentication to CrowdStrike API
    // POST https://api.crowdstrike.com/oauth2/token
    
    // TODO: Implement actual API call
    // For now, simulate connection
    connected_ = false;
    
    if (connected_) {
        std::cout << "[+] Connected to CrowdStrike" << std::endl;
    }
    
    return connected_;
}

void CrowdStrikeConnector::disconnect() {
    connected_ = false;
}

EDRStatus CrowdStrikeConnector::getStatus() {
    EDRStatus status;
    status.type = EDRType::CROWDSTRIKE;
    status.name = "CrowdStrike Falcon";
    status.active = connected_;
    status.realTimeProtection = true;
    return status;
}

std::vector<std::string> CrowdStrikeConnector::getAlerts(int count) {
    std::vector<std::string> alerts;
    // GET /alerts/queries/alerts/v1
    return alerts;
}

bool CrowdStrikeConnector::isConnected() const {
    return connected_;
}

// ============================================================================
// MICROSOFT DEFENDER CONNECTOR
// ============================================================================

DefenderConnector::DefenderConnector(const std::string& tenantId,
                                      const std::string& clientId,
                                      const std::string& clientSecret)
    : tenantId_(tenantId)
    , clientId_(clientId)
    , clientSecret_(clientSecret)
    , connected_(false) {
}

DefenderConnector::~DefenderConnector() {
    disconnect();
}

bool DefenderConnector::connect() {
    std::cout << "[*] Connecting to Microsoft Defender ATP API..." << std::endl;
    
    // OAuth2 to Azure AD
    // POST https://login.microsoftonline.com/{tenantId}/oauth2/token
    
    connected_ = false;
    return connected_;
}

void DefenderConnector::disconnect() {
    connected_ = false;
}

EDRStatus DefenderConnector::getStatus() {
    EDRStatus status;
    status.type = EDRType::DEFENDER;
    status.name = "Microsoft Defender";
    status.active = isDefenderActive();
    status.realTimeProtection = isRealTimeProtectionEnabled();
    return status;
}

std::vector<std::string> DefenderConnector::getAlerts(int count) {
    std::vector<std::string> alerts;
    // GET /api/alerts
    return alerts;
}

bool DefenderConnector::isConnected() const {
    return connected_;
}

bool DefenderConnector::isDefenderActive() {
#ifdef _WIN32
    // Check if MsMpEng.exe is running
    // Or use WMI: SELECT * FROM MSFT_MpComputerStatus
    
    SC_HANDLE scm = OpenSCManager(NULL, NULL, SC_MANAGER_CONNECT);
    if (scm) {
        SC_HANDLE svc = OpenService(scm, L"WinDefend", SERVICE_QUERY_STATUS);
        if (svc) {
            SERVICE_STATUS status;
            if (QueryServiceStatus(svc, &status)) {
                CloseServiceHandle(svc);
                CloseServiceHandle(scm);
                return status.dwCurrentState == SERVICE_RUNNING;
            }
            CloseServiceHandle(svc);
        }
        CloseServiceHandle(scm);
    }
#endif
    return false;
}

bool DefenderConnector::isRealTimeProtectionEnabled() {
#ifdef _WIN32
    // Query MSFT_MpPreference for RealTimeProtectionEnabled
    // Using WMI or registry
    HKEY hKey;
    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE,
                      L"SOFTWARE\\Microsoft\\Windows Defender\\Real-Time Protection",
                      0, KEY_READ, &hKey) == ERROR_SUCCESS) {
        DWORD value = 0;
        DWORD size = sizeof(value);
        if (RegQueryValueExW(hKey, L"DisableRealtimeMonitoring", NULL, NULL,
                             (LPBYTE)&value, &size) == ERROR_SUCCESS) {
            RegCloseKey(hKey);
            return value == 0;  // 0 means RTP is enabled
        }
        RegCloseKey(hKey);
    }
#endif
    return false;
}

// ============================================================================
// CARBON BLACK CONNECTOR
// ============================================================================

CarbonBlackConnector::CarbonBlackConnector(const std::string& apiUrl,
                                            const std::string& apiKey,
                                            const std::string& orgKey)
    : apiUrl_(apiUrl)
    , apiKey_(apiKey)
    , orgKey_(orgKey)
    , connected_(false) {
}

CarbonBlackConnector::~CarbonBlackConnector() {
    disconnect();
}

bool CarbonBlackConnector::connect() {
    std::cout << "[*] Connecting to Carbon Black API..." << std::endl;
    
    // API authentication
    // X-Auth-Token header with API key
    
    connected_ = false;
    return connected_;
}

void CarbonBlackConnector::disconnect() {
    connected_ = false;
}

EDRStatus CarbonBlackConnector::getStatus() {
    EDRStatus status;
    status.type = EDRType::CARBONBLACK;
    status.name = "VMware Carbon Black";
    status.active = connected_;
    return status;
}

std::vector<std::string> CarbonBlackConnector::getAlerts(int count) {
    std::vector<std::string> alerts;
    return alerts;
}

bool CarbonBlackConnector::isConnected() const {
    return connected_;
}

} // namespace integration
} // namespace edr
